src/Storages/MergeTree/MergeTreeSplitPrewhereIntoReadSteps.cpp

```diff
@@ -146,6 +149,27 @@ const ActionsDAG::Node & addCast(
     return addFunction(dag, func_builder_cast, std::move(children), node_remap);
 }
 
+/// Normalizes the filter node by adding AND with a constant true.
+/// This:
+/// 1. produces a result with the proper Nullable or non-Nullable UInt8 type and
+/// 2. makes sure that the result contains only 0 or 1 values even if the source column contains non-boolean values.
```

Comment:
why it is necessary?

Reply:
It happens if we have a table (x Nullable(UInt8), y UInt8) with values
1 1
2 2
3 3
and run a query:
```SELECT x AND y PREWHERE x AND y;```
In the original DAG PREWHERE column is ```and(x, y)``` and has type Nullable(UInt8) and values only 0, 1 or Null. This columns is computed in PREWHERE and then returned by SELECT, so the result is
1
1
1
When we split PREWHERE DAG into steps the second step will return just ```y``` which in non-Nullable and contains 1,2,3. So in order to match the original output we and it with True to make it only 0s and 1s, and cast to Nullable(Uint8)

Reply:
Also 'AND true' is needed if the original DAG contained something LIKE ```x AND y``` and x,y are both UInt64. They need to be cast to Bool to serve as filters in the individual steps.

Reply:
I mean is it really necessary for the result column to not have values > 1? just `CAST(and_node, type of PREWHERE column)` isn't enough?

Reply:
Yep, in the example above ```SELECT x AND y PREWHERE x AND y;``` will return
1
2
3
if we just cast last step filter column ```y``` to Uint8, but the correct result is
1
1
1